internal/server/providers/azure_test.go

```diff
@@ -0,0 +1,328 @@
+			"onPremisesLastSyncDateTime": null,
+			"onPremisesNetBiosName": null,
+			"onPremisesSamAccountName": null,
+			"onPremisesSecurityIdentifier": null,
+			"onPremisesSyncEnabled": null,
+			"preferredDataLocation": null,
+			"preferredLanguage": null,
+			"proxyAddresses": [],
+			"renewedDateTime": "2022-06-07T20:40:20Z",
+			"resourceBehaviorOptions": [],
+			"resourceProvisioningOptions": [],
+			"securityEnabled": true,
+			"securityIdentifier": "qqq",
+			"theme": null,
+			"visibility": null,
+			"onPremisesProvisioningErrors": []
+		}
+	]
+}`
+
+func TestSyncAzureProviderUser(t *testing.T) {
+	server, ctx := setupOIDCTest(t)
+	serverURL := server.run(t)
+
+	db := setupDB(t)
+
+	provider := &models.Provider{
+		Name: "mock-azure",
+		Kind: models.AzureKind,
+	}
+
+	err := data.CreateProvider(db, provider)
+	assert.NilError(t, err)
+
+	oidc := &oidcImplementation{
+		ProviderID:   provider.ID,
+		Domain:       serverURL,
+		ClientID:     "whatever",
+		ClientSecret: "whatever",
+		RedirectURL:  "http://localhost:8301",
+	}
+	azure := &azure{OIDC: oidc}
+
+	now := time.Now().UTC()
+
+	claims := jwt.Claims{
+		Audience:  jwt.Audience([]string{server.knownClientID}),
+		NotBefore: jwt.NewNumericDate(now.Add(time.Minute * -5)), // adjust for clock drift
+		Expiry:    jwt.NewNumericDate(now.Add(time.Minute * 5)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		Issuer:    serverURL,
+	}
+
+	body, err := testTokenResponse(claims, server.signingKey, "hello@example.com")
+	assert.NilError(t, err)
+
+	server.tokenResponse = tokenResponse{
+		code: 200,
+		body: body,
+	}
+
+	tests := []struct {
+		name         string
+		setupFunc    func(t *testing.T) *models.Identity
+		infoResponse string
+		verifyFunc   func(t *testing.T, err error, user *models.Identity)
+	}{
+		{
+			name: "invalid/expired access token is updated",
+			setupFunc: func(t *testing.T) *models.Identity {
+				user := &models.Identity{
+					Name: "sharrington@example.com",
+				}
+
+				err = data.CreateIdentity(db, user)
+				assert.NilError(t, err)
+
+				pu := &models.ProviderUser{
+					ProviderID: provider.ID,
+					IdentityID: user.ID,
+
+					Email:        user.Name,
+					RedirectURL:  "http://example.com",
+					AccessToken:  models.EncryptedAtRest("aaa"),
+					RefreshToken: models.EncryptedAtRest("bbb"),
+					ExpiresAt:    time.Now().Add(time.Minute * -5),
+					LastUpdate:   time.Now(),
+				}
+
+				err = data.UpdateProviderUser(db, pu)
+				assert.NilError(t, err)
+
+				return user
+			},
+			infoResponse: `{
+				"sub": "o_aaabbbccc",
+				"name": "Steve Harrington",
+				"family_name": "Harrington",
+				"given_name": "Steve",
+				"picture": "https://graph.microsoft.com/v1.0/me/photo/$value"
+			}`,
+			verifyFunc: func(t *testing.T, err error, user *models.Identity) {
+				assert.NilError(t, err)
+
+				pu, err := data.GetProviderUser(db, provider.ID, user.ID)
+				assert.Assert(t, err)
+
+				assert.Assert(t, pu.AccessToken != "aaa")
```

Comment:
Are there other `ProviderUser` fields that are modified that we should check here? Should `ExpiredAt` be updated as well? `Groups` ? `LastUpdated` ?

Check out `cmpClientHostConfig` for an example of how we can customize a comparison. You can set time thresholds so that checking a `time.Time` doesn't fail if it's only a few seconds different.  That gets passed to `assert.DeepEqual`, so that we can check the whole structure is updated as expected.

Reply:
Fixed now from your example, doing the deep equal compares